akka-stream/src/main/scala/akka/stream/impl/io/SslTls.scala

```diff
@@ -0,0 +1,419 @@
+    }
+    sslParameters foreach (p ⇒ engine.setSSLParameters(p))
+    engine.beginHandshake()
+    lastHandshakeStatus = engine.getHandshakeStatus
+  }
+
+  def setNewSessionParameters(n: NegotiateNewSession): Unit = {
+    if (tracing) log.debug(s"applying $n")
+    currentSession.invalidate()
+    currentSessionParameters = n
+    applySessionParameters()
+    corkUser = true
+  }
+
+  /*
+   * So here’s the big picture summary: the SSLEngine is the boss, and it can
+   * be in several states. Depending on this state, we may want to react to
+   * different input and output conditions.
+   *
+   *  - normal bidirectional operation (does both outbound and inbound)
+   *  - outbound close initiated, inbound still open
+   *  - inbound close initiated, outbound still open
+   *  - fully closed
+   *
+   * Upon reaching the last state we obviously just shut down. In addition to
+   * these user-data states, the engine may at any point in time also be
+   * handshaking. This is mostly transparent, but it has an influence on the
+   * outbound direction:
+   *
+   *  - if the local user triggered a re-negotiation, cork all user data until
+   *    that is finished
+   *  - if the outbound direction has been closed, trigger outbound readiness
+   *    based upon HandshakeStatus.NEED_WRAP
+   *
+   * These conditions lead to the introduction of a synthetic TransferState
+   * representing the Engine.
+   */
+
+  var lastHandshakeStatus: HandshakeStatus = _
+
+  val engineNeedsWrap = new TransferState {
+    def isReady = lastHandshakeStatus == NEED_WRAP
+    def isCompleted = false
+  }
+
+  val engineInboundOpen = new TransferState {
+    def isReady = !engine.isInboundDone()
+    def isCompleted = false
+  }
+
+  var corkUser = true
+
+  val userHasData = new TransferState {
+    private val user = inputBunch.inputsOrCompleteAvailableFor(UserIn) || userInChoppingBlock
+    def isReady = !corkUser && user.isReady && lastHandshakeStatus != NEED_UNWRAP
+    def isCompleted = false
+  }
+
+  val transportHasData = inputBunch.inputsOrCompleteAvailableFor(TransportIn) || transportInChoppingBlock
+
+  // bidirectional case
+  val outbound = (userHasData || engineNeedsWrap) && outputBunch.demandAvailableFor(TransportOut)
+  val inbound = transportHasData && outputBunch.demandOrCancelAvailableFor(UserOut)
+
+  // half-closed
+  val outboundHalfClosed = engineNeedsWrap && outputBunch.demandAvailableFor(TransportOut)
+  val inboundHalfClosed = transportHasData && engineInboundOpen
+
+  def completeOrFlush(): Unit =
+    if (engine.isOutboundDone()) nextPhase(completedPhase)
+    else nextPhase(flushingOutbound)
+
+  private def doInbound(isOutboundClosed: Boolean, inboundState: TransferState): Boolean =
+    if (inputBunch.isDepleted(TransportIn) && transportInChoppingBlock.isEmpty) {
+      if (tracing) log.debug("closing inbound")
+      try engine.closeInbound()
+      catch { case ex: SSLException ⇒ outputBunch.enqueue(UserOut, SessionTruncated) }
```

Comment:
This is explained in the JavaDoc of `closeInbound`, so I think this is sufficiently clear.